### PR TITLE
EZEE-3159: Added support for items in the trash for SubtreeLimitationType

### PR DESCRIPTION
**Pull request by mateuszdebinski**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZEE-3159](https://jira.ez.no/browse/EZEE-3159)
| **Bug**| yes
| **New feature**    | no
| **Target version** | `7.5` 
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | no

**TODO**:
- [x] Implement feature / fix a bug.
- [x] Implement tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

File: eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway.php [hunk 1]
```diff
@@ -76,6 +76,17 @@ abstract class Gateway
      */
     abstract public function loadLocationDataByContent($contentId, $rootLocationId = null);
 
+    /**
+     * Loads data for all Locations for $contentId in trash, optionally only in the
+     * subtree starting at $rootLocationId.
+     *
+     * @param int $contentId
+     * @param int $rootLocationId
```
Comment (mikadamczyk) [inline eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway.php]:
```suggestion
     * @param int $rootLocationId|null
```

File: eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway.php [hunk 2]
```diff
@@ -76,6 +76,17 @@ abstract class Gateway
...
+     *
+     * @return array
+     */
+    abstract public function loadLocationDataByTrashContent($contentId, $rootLocationId = null);
```
Comment (mikadamczyk) [inline eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway.php]: This is a new method and we can add type hint here

File: eZ/Publish/SPI/Persistence/Content/Location/Handler.php [hunk 3]
```diff
@@ -75,6 +75,14 @@ public function loadByRemoteId($remoteId, array $translations = null, bool $useA
      */
     public function loadLocationsByContent($contentId, $rootLocationId = null);
 
+    /**
+     * Loads all locations for $contentId in trash, optionally limited to a sub tree
```
Comment (ViniTou) [inline eZ/Publish/SPI/Persistence/Content/Location/Handler.php:79]: Not a fan of this comment repeated 4 times.

File: eZ/Publish/Core/Persistence/Legacy/Content/Location/Mapper.php [hunk 4]
```diff
@@ -56,17 +56,18 @@ public function createLocationFromRow(array $data, $prefix = '', Location $locat
      *
      * @param array $rows
      * @param string $prefix
+     * @param Location $location
```
Comment (ViniTou) [inline eZ/Publish/Core/Persistence/Legacy/Content/Location/Mapper.php]: FQCN

File: eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway.php [hunk 5]
```diff
@@ -70,12 +70,18 @@ abstract class Gateway
...
      *
      * @return array
      */
     abstract public function loadLocationDataByContent($contentId, $rootLocationId = null);
 
+    /**
+     * Loads data for all Locations for $contentId in trash, optionally only in the
+     * subtree starting at $rootLocationId.
+     */
+    abstract public function loadLocationDataByTrashContent(int $contentId, int $rootLocationId = null): array;
```
Comment (ViniTou) [inline eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway.php]: typehint specific array in phpdoc

**Review by ViniTou — COMMENTED** [review]
Are we still waiting for unit/integration tests?

File: eZ/Publish/Core/Persistence/Cache/LocationHandler.php [hunk 6]
```diff
@@ -145,6 +145,16 @@ public function loadLocationsByContent($contentId, $rootLocationId = null)
         return $locations;
     }
 
+    /**
+     * {@inheritdoc}
+     */
+    public function loadLocationsByTrashContent(int $contentId, int $rootLocationId = null): array
```
Comment (mikadamczyk) [inline eZ/Publish/Core/Persistence/Cache/LocationHandler.php]:
```suggestion
    public function loadLocationsByTrashContent(int $contentId, ?int $rootLocationId = null): array
```

File: eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway.php [hunk 7]
```diff
@@ -70,12 +70,20 @@ abstract class Gateway
...
      */
     abstract public function loadLocationDataByContent($contentId, $rootLocationId = null);
 
+    /**
+     * Loads data for all Locations for $contentId in trash, optionally only in the
+     * subtree starting at $rootLocationId.
+     *
+     * @return string[]
+     */
+    abstract public function loadLocationDataByTrashContent(int $contentId, int $rootLocationId = null): array;
```
Comment (mikadamczyk) [inline eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway.php]:
```suggestion
    abstract public function loadLocationDataByTrashContent(int $contentId, ?int $rootLocationId = null): array;
```

File: eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway/DoctrineDatabase.php [hunk 8]
```diff
@@ -136,17 +136,40 @@ public function loadLocationDataByContent($contentId, $rootLocationId = null)
...
         $statement = $query->prepare();
         $statement->execute();
 
         return $statement->fetchAll(\PDO::FETCH_ASSOC);
     }
 
+    /**
+     * {@inheritdoc}
+     */
+    public function loadLocationDataByTrashContent(int $contentId, int $rootLocationId = null): array
```
Comment (mikadamczyk) [inline eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway/DoctrineDatabase.php]:
```suggestion
    public function loadLocationDataByTrashContent(int $contentId, ?int $rootLocationId = null): array
```

File: eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway/ExceptionConversion.php [hunk 9]
```diff
@@ -71,6 +71,15 @@ public function loadLocationDataByContent($contentId, $rootLocationId = null)
         }
     }
 
+    public function loadLocationDataByTrashContent(int $contentId, int $rootLocationId = null): array
```
Comment (mikadamczyk) [inline eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway/ExceptionConversion.php]:
```suggestion
    public function loadLocationDataByTrashContent(int $contentId, ?int $rootLocationId = null): array
```

File: eZ/Publish/Core/Persistence/Legacy/Content/Location/Handler.php [hunk 10]
```diff
@@ -159,6 +160,16 @@ public function loadLocationsByContent($contentId, $rootLocationId = null)
         return $this->locationMapper->createLocationsFromRows($rows);
     }
 
+    /**
+     * {@inheritdoc}
+     */
+    public function loadLocationsByTrashContent(int $contentId, int $rootLocationId = null): array
```
Comment (mikadamczyk) [inline eZ/Publish/Core/Persistence/Legacy/Content/Location/Handler.php]:
```suggestion
    public function loadLocationsByTrashContent(int $contentId, ?int $rootLocationId = null): array
```

File: eZ/Publish/Core/Persistence/Legacy/Content/Location/Mapper.php [hunk 11]
```diff
@@ -56,17 +56,18 @@ public function createLocationFromRow(array $data, $prefix = '', Location $locat
      *
      * @param array $rows
      * @param string $prefix
+     * @param \eZ\Publish\SPI\Persistence\Content\Location $location
      *
      * @return \eZ\Publish\SPI\Persistence\Content\Location[]
      */
-    public function createLocationsFromRows(array $rows, $prefix = '')
+    public function createLocationsFromRows(array $rows, $prefix = '', Location $location = null)
```
Comment (mikadamczyk) [inline eZ/Publish/Core/Persistence/Legacy/Content/Location/Mapper.php]:
```suggestion
    public function createLocationsFromRows(array $rows, $prefix = '', ?Location $location = null)
```

File: eZ/Publish/SPI/Persistence/Content/Location/Handler.php [hunk 12]
```diff
@@ -75,6 +75,14 @@ public function loadByRemoteId($remoteId, array $translations = null, bool $useA
...
+     * identified by $rootLocationId.
+     *
+     * @return \eZ\Publish\SPI\Persistence\Content\Location[]
+     */
+    public function loadLocationsByTrashContent(int $contentId, int $rootLocationId = null): array;
```
Comment (mikadamczyk) [inline eZ/Publish/SPI/Persistence/Content/Location/Handler.php]:
```suggestion
    public function loadLocationsByTrashContent(int $contentId, ?int $rootLocationId = null): array;
```

**Review by mikadamczyk — APPROVED** [review]

**Review by konradoboza — APPROVED** [review]

File: eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway/DoctrineDatabase.php [hunk 13]
```diff
@@ -229,12 +252,13 @@ public function getSubtreeContent($sourceId, $onlyIds = false)
      *
      * @param \eZ\Publish\Core\Persistence\Database\Query $query
      * @param string $rootLocationId
+     * @param string $tableName
      */
-    protected function applySubtreeLimitation(DatabaseQuery $query, $rootLocationId)
+    protected function applySubtreeLimitation(DatabaseQuery $query, $rootLocationId, string $tableName = 'ezcontentobject_tree')
```
Comment (devmysz) [inline eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway/DoctrineDatabase.php]: Please split this line and add return type `void`

File: eZ/Publish/Core/Persistence/Legacy/Content/Location/Mapper.php [hunk 14]
```diff
@@ -23,11 +23,11 @@ class Mapper
      *
      * @param array $data
      * @param string $prefix
-     * @param Location $location
+     * @param \eZ\Publish\SPI\Persistence\Content\Location $location
```
Comment (devmysz) [inline eZ/Publish/Core/Persistence/Legacy/Content/Location/Mapper.php]:
```suggestion
     * @param \eZ\Publish\SPI\Persistence\Content\Location|null $location
```

File: eZ/Publish/Core/Persistence/Legacy/Content/Location/Mapper.php [hunk 15]
```diff
@@ -56,17 +56,18 @@ public function createLocationFromRow(array $data, $prefix = '', Location $locat
      *
      * @param array $rows
      * @param string $prefix
+     * @param \eZ\Publish\SPI\Persistence\Content\Location $location
```
Comment (devmysz) [inline eZ/Publish/Core/Persistence/Legacy/Content/Location/Mapper.php]:
```suggestion
     * @param \eZ\Publish\SPI\Persistence\Content\Location|null $location
```

File: eZ/Publish/API/Repository/Tests/TrashServiceTest.php [hunk 16]
```diff
@@ -801,6 +801,66 @@ public function testFindTrashItemsLimitedAccess()
...
+        $user = $this->createCustomUserWithLogin(
+            'user',
+            'user@example.com',
+            'roleTrashCleaners',
+            'roleTrashCleaner'
+        );
+        $repository->getPermissionResolver()->setCurrentUserReference($user);
+
+        $query = new Query();
+        $query->filter = new Criterion\LogicalAnd(
```
Comment (devmysz) [inline eZ/Publish/API/Repository/Tests/TrashServiceTest.php]: This filter is unnecesary if you want to get all trash items

**Review by devmysz — CHANGES_REQUESTED** [review]

**Review by devmysz — APPROVED** [review]
👌

**Review by adamwojs — APPROVED** [review]

**Review by katarzynazawada — APPROVED** [review]

**Comment by lserwatka** [comment]
@mateuszbieniek you can merge it up to ezplatform-kernel/1.0 and master.

**Comment by mateuszbieniek** [comment]
@lserwatka I think you pinged wrong Mateusz ;) @mateuszdebinski

**Comment by mateuszdebinski** [comment]
@lserwatka I did it in the morning

**Pull request merged**